### ISSUE TITLE
CSS files should be included in the build

### ISSUE THREAD

**Issue by janjansky** [ngbp/ngbp#57]
I am using several css files for my site made by myself.
not less file. pure CSS file.
And I know the tool of converting CSS to less.

But I want to use pure CSS files, without converting to less.
How to import them?

**Comment by sudhakar** [comment]
Since `less` files are simply a superset of `css`, one possible solution would be rename your `css` with `less` extension and simply import them into `main.less` file.

**Comment by janjansky** [comment]
I added custom css file to `build.config.js` file.
But after building (`grunt watch`), in `vendor` directory, I can not find these files.

``` javascript
vendor_files: {
    js: [
     ....
    ],
    css: [
        'vendor/additional/css/select.css',
        'vendor/additional/css/main.css'
    ]
```

**Comment by janjansky** [comment]
Thank you for your reply.
I think your way is one of ways on how to solve this problem.

`Important question`
After bower install, I can not look for `bootstrap.editable.less` in `vendor/bootstrap`.
In order to solve this problem, how and what do I add to `Gruntfile.js` and `bower.json`?

**Comment by joshdmiller** [comment]
@janjansky - The issue with the vendor files seems to be that I never implemented that. Oops. They were supposed to concatenate with the compiled LESS sources but I can see that code is nowhere to be found. Would you mind opening a separate issue for that? If you do, I'll provide a patch you can use until the new version of ngBoilerplate is released; it's really a rather simple change.

As for using CSS, any files that you put in `src/assets/` will be automatically copied to the resulting build, so you if you have just a single CSS file, you can place it in assets and add it to `index.html` manually. There of course would be no watching or concatenation, which is totally lame. So we need to fix this! :-)

I think the best way to fix this is to implement this feature with the SASS one; in essence, a three-way toggle: CSS, LESS, or SASS. If CSS, then the CSS files are linted and concatenated, but if one of the latter then they are appropriately compiled as well. I'm going to change the title of this issue to be "Support Plain CSS" and mark it as an enhancement for the upcoming 0.4 version.

How does all that sound?

**Comment by joshdmiller** [comment]
@janjansky - Your second reply was not visible when I posted this; I'm on  mobile using an app that doesn't auto-refresh. What is `bootstrap.editable.less`? The LESS versions are stored in `vendor/bootstrap/less` and the overrides are stored in `src/less`.

**Comment by janjansky** [comment]
Thank you. Really thank you. @joshdmiller 
First of all, I agree with your opinion of changing the title of issue.

Second.
I understand 2 ways suggested by you.

Third.
As I mentioned above, to build my css files, I added them to `build.config.js`.
But after building (`grunt build`), I can not find these files in `build/vender/mydirecotry/css/`.

Thank you again.

**Comment by janjansky** [comment]
In index.html of your project, in order to import complied CSS files, you described as follows:

``` html
    <!-- compiled CSS --><% styles.forEach( function ( file ) { %>
    <link rel="stylesheet" type="text/css" href="<%= file %>" /><% }); %>
```

But as I added my css files to `build.config.js` file, all of which are in `vendor` directory, they are not complied.
If you are possible, please reply me.

**Comment by joshdmiller** [comment]
@janjansky As I said in my previous comment, there is no current functionality to include css files in the build. The listed vendor files _should_ be included, but they are not due to a bug.

**Comment by grantlucas** [comment]
Created a new issue (#113) before I remembered this one existed.

I guess the feature request is while adding Vendor CSS support, there should be a way to specify IE versions if needed.

**Comment by joshdmiller** [comment]
Closed via dda8792.

**Comment by tacticiankerala** [comment]
I recently cloned this and was playing around with angular a little bit and I ran into the same issue. It was surprising to see that the problem has been fixed in an earlier version but still I was facing the issue.

After the fix has been issued, I think a new commit ( https://github.com/ngbp/ng-boilerplate/commit/29502bff8eb649501945aef37fe50d81c36151ba ) has been added which brought back the issue  (I think it did. Please feel free to correct me if I am missing something).

I was playing around with the Gruntfile a little bit and I found adding vendor_file.css into the recess:build has solved my issue. 

``` javascript
recess: {
      build: {
        src: [ '<%= app_files.less %>','<%= vendor_files.css %>' ],
```

Please let me know if I am doing something wrong.

**Comment by joshdmiller** [comment]
It should still be working fine. It's not needed in the `recess` task because it doesn't need to pass through the less compiler. Instead, there's a concat task for it: https://github.com/ngbp/ng-boilerplate/blob/v0.3.1-release/Gruntfile.js#L164.

Can you post your `vendor_files` object?

**Comment by tacticiankerala** [comment]
Please find the vendor_files and Gruntfile configuration below,

vendor_files  - https://gist.github.com/tacticiankerala/6618569 
Gruntfile.js - https://gist.github.com/tacticiankerala/6618584

**Comment by bf0** [comment]
The copy code mentioned is there and it works, sort of. The issue has to do with the index file still having:

```
<!-- compiled CSS --><% styles.forEach( function ( file ) { %>
<link rel="stylesheet" type="text/css" href="<%= file %>" /><% }); %>
```

... which picks up vendor_files.css

so now the vendor_files css entries are appended to the less file, great, but the rendered index.html is now asking for the non-appended file it is expecting in the vendor path. Since it is appended and not copied, index.html doesn't find it, and throws an error.

as far as I can see, you cannot add a file to vendor_files.css without it causing a resource not found error - or am I missing something?

**Comment by cr0cK** [comment]
Imho, you're right. I've got the same problem.

**Comment by eladmoshe** [comment]
@bf0 I agree, same here.

**Comment by scribblet** [comment]
Same problem here.

**Comment by flocks** [comment]
I had more or less the same problem. It worked well the first time I ran "grunt watch" ( concat the css vendors with the compiled css ), but every time I was modifying a less file, it ran the less task without concatening the css vendors.

So first, I delete the line referencing "vendor_files.css" in index['build']['src'] in order to prevent the index.html to include each vendor css file.

Then I added the task concat:build_css into the less task 

```
  less: {
    files: [ 'src/**/*.less' ],
    tasks: [ 'less:build' , 'concat:build_css']
  },
```

It worked for me, but not sure if it is the proper way to do it.

**Comment by czosel** [comment]
@flocks: Exactly the same problem here. Your fix looks perfect to me, thanks!

**Comment by ladaltamirano** [comment]
Same here...

**Comment by yveoch** [comment]
same

**Comment by rasekar2000** [comment]
In the Line 170 , I have added less: above the build_css .. Still I am getting the same error.. 

<% styles.forEach( function ( file ) { %> <% }); %> <% scripts.forEach( function ( file ) { %> <% }); %> 

/**
     * `grunt concat` concatenates multiple source files into a single file.
     */
    concat: {
      
      /** 
      https://github.com/ngbp/ngbp/issues/57#issuecomment-45731609
      */
      less: {
        files: [ 'src/**/*.less' ],
        tasks: [ 'less:build' , 'concat:build_css']
      },
      /**
       * The `build_css` target concatenates compiled CSS and vendor CSS
       * together.
       */
      build_css: {
        src: [
          '<%= vendor_files.css %>',
          '<%= build_dir %>/assets/<%= pkg.name %>-<%= pkg.version %>.css'
        ],
        dest: '<%= build_dir %>/assets/<%= pkg.name %>-<%= pkg.version %>.css'
      },

**Comment by AndyBrownlie** [comment]
@flocks: I'm having this same problem with select.css from ui-select.  When I try to adjust any other styling, I then get strange styling behaviour with the select dropdown (textbox showing directly below or inside the dropdown).

I tried your suggestion too, I'm assuming the changes you are referring to are on the gruntfile.js?
When I do this, it includes the select css styling in the overall <project>.css but it still renders the textbox styling referred to above. 

I know for a fact that this stylesheet is the problem because when I manually add it, all styling issues disappear, yet I cannot seem to get it working using your suggested workaround.

@flocks & @joshdmiller, do you have any suggestions for this?

**Comment by joshdmiller** [comment]
@AndyBrownlie This is issue is closed as resolved through a code push. As far as I can tell, vendor CSS is included in the build. Is this a different problem, an edge case the solution doesn't cover, or confirmation that the fix doesn't work _generally_?

A reproducible case would be super helpful.

**Comment by AndyBrownlie** [comment]
Thanks for the reply @joshdmiller. I'm happy to try to reproduce, but I'm struggling to think how. Its not something I can do in plunker/jsfiddle as I think its to do with the framework itself rather than individual code elements (as I say, I can get rid of the issue by manually copying the select.css file into the build/vendor/ui-select/dist folder so this suggests its something to do with the build files). Any advice is welcomed.

What I can't understand is that I definitely can get the styling from the select.css concatenated into the overall <project>.css, but for some reason I'm still getting the following:

![screen shot 2015-01-29 at 08 42 24](https://cloud.githubusercontent.com/assets/857089/5952555/fcc4b02c-a792-11e4-9044-b16209aeef86.png)

**Comment by joshdmiller** [comment]
@AndyBrownlie I more meant steps to reproduce. E.g.:
1. get a fresh copy of ng-boilerplate
2. require a vendor css file
3. run the build
4. ... ?
5. instead of [?], I see [?]

Then I can see what's happening.

**Comment by AndyBrownlie** [comment]
Oh right, understood.

Well, I downloaded ng-boilerplate, then some time later added the ui-select
code via bower install.  (I am using the bootstrap theme on the select
dropdowns.)
I noticed immediately that the file was missing from the build/vendor dir
and producing an error when looking at the console window in the Chrome dev
tools. However, it didn't seem to be adversely affecting any of the styling
at that point.
Its only when I then wanted to add some custom styling did I then get the
style issues (captured and included in my last post).
When I manually add the select.css file after a grunt build the styling
issues immediately disappear.
As I previously said, I followed the concat instructions, and it seemed to
be working (I checked the <project>.css file and it contained the
select.css styling and yet it didn't seem to eradicate the issue.  Could it
be an issue withe the ui-select code/build?

I'm using angular-ui-select 0.8.3, angular 1.2, angular-boostrap 0.12.0 and
ngbp 0.3.1 (I think).

Any help is gratefully received.

regards

Andy Brownlie
[image: http://]about.me/andybrownlie
http://about.me/andybrownlie?promo=email_sig
